CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+  Prefer `{...}` over `do...end` for single-line blocks.  Avoid using `{...}`
+  for multi-line blocks (multiline chaining is always ugly). Always use
+  `do...end` for "control flow" and "method definitions" (e.g. in Rakefiles and
+  certain DSLs).  Avoid `do...end` when chaining.
+<sup>[[link](#single-line-blocks)]</sup>
+
+  ```Ruby
+  names = %w(Bozhidar Steve Sarah)
+
+  # bad
+  names.each do |name|
+    puts name
+  end
+
+  # good
+  names.each { |name| puts name }
+
+  # bad
+  names.select do |name|
+    name.start_with?('S')
+  end.map { |name| name.upcase }
+
+  # good
+  names.select { |name| name.start_with?('S') }.map(&:upcase)
+  ```
+
+  Some will argue that multiline chaining would look OK with the use of {...},
+  but they should ask themselves - is this code really readable and can the
+  blocks' contents be extracted into nifty methods?
+
+* <a name="block-argument"></a>
+  Consider using explicit block argument to avoid writing block literal that
+  just passes its arguments to another block. Beware of the performance impact,
+  though, as the block gets converted to a Proc.
+<sup>[[link](#block-argument)]</sup>
+
+  ```Ruby
+  require 'tempfile'
+
+  # bad
+  def with_tmp_dir
+    Dir.mktmpdir do |tmp_dir|
+      Dir.chdir(tmp_dir) { |dir| yield dir }  # block just passes arguments
+    end
+  end
+
+  # good
+  def with_tmp_dir(&block)
+    Dir.mktmpdir do |tmp_dir|
+      Dir.chdir(tmp_dir, &block)
+    end
+  end
+
+  with_tmp_dir do |dir|
+    puts "dir is accessible as a parameter and pwd is set: #{dir}"
+  end
+  ```
+
+* <a name="no-explicit-return"></a>
+  Avoid `return` where not required for flow of control.
+<sup>[[link](#no-explicit-return)]</sup>
+
+  ```Ruby
+  # bad
+  def some_method(some_arr)
+    return some_arr.size
+  end
+
+  # good
+  def some_method(some_arr)
+    some_arr.size
+  end
+  ```
+
+* <a name="no-self-unless-required"></a>
+  Avoid `self` where not required. (It is only required when calling a self
+  write accessor.)
+<sup>[[link](#no-self-unless-required)]</sup>
+
+  ```Ruby
+  # bad
+  def ready?
+    if self.last_reviewed_at > self.last_updated_at
+      self.worker.update(self.content, self.options)
+      self.status = :in_progress
+    end
+    self.status == :verified
+  end
+
+  # good
+  def ready?
+    if last_reviewed_at > last_updated_at
+      worker.update(content, options)
+      self.status = :in_progress
+    end
+    status == :verified
+  end
+  ```
+
+* <a name="no-shadowing"></a>
+  As a corollary, avoid shadowing methods with local variables unless they are
+  both equivalent.
+<sup>[[link](#no-shadowing)]</sup>
+
+  ```Ruby
+  class Foo
+    attr_accessor :options
+
+    # ok
+    def initialize(options)
+      self.options = options
+      # both options and self.options are equivalent here
+    end
+
+    # bad
+    def do_something(options = {})
+      unless options[:when] == :later
+        output(self.options[:message])
+      end
+    end
+
+    # good
```

Comment:
+1 on avoiding shadowing even between scopes (local/method arg of `options` conflicting with the instance method `self.options`). Clarity matters, so let's try to avoid having a local var shadowing a method/etc.